src/librustc_resolve/lib.rs

```diff
@@ -883,6 +903,16 @@ impl<'a, 'tcx> Visitor<'tcx> for Resolver<'a> {
                 }
             }));
 
+        // We also ban access to type parameters for use as the types of const parameters.
+        let mut const_ty_param_ban_rib = Rib::new(TyParamAsConstParamTy);
+        const_ty_param_ban_rib.bindings.extend(generics.params.iter()
```

Comment:
I'm not sure this is something that needs to be done in resolve.
A parameter type is just a special case of non-"structural match" types that need to be rejected somewhere else.

Reply:
It's possible that we'll want to accept type parameter as types for const parameters at some point in the future, which is why I deliberately special-cased this here, so it's explicit, rather than falling out of some other checks. The other checks in resolve looked like a similar style of error, but perhaps there's a better place to put it?

Reply:
Special-case the non-"structural match" message if the type is a type parameter?

Reply:
Okay, I'll move this error to a later PR.